src/mono/mono/mini/mini-llvm.c

```diff
@@ -5993,10 +6001,33 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			case LLVMArgAsIArgs:
 			case LLVMArgFpStruct: {
 				LLVMTypeRef ret_type = LLVMGetReturnType (LLVMGetElementType (LLVMTypeOf (method)));
-				LLVMValueRef retval;
-
-				g_assert (addresses [ins->sreg1]);
-				retval = LLVMBuildLoad2 (builder, ret_type, convert (ctx, addresses [ins->sreg1]->value, pointer_type (ret_type)), "");
+				LLVMValueRef retval, elem;
+				MonoClass *klass = mono_class_from_mono_type_internal (sig->ret);
+				gboolean is_simd = MONO_CLASS_IS_SIMD (ctx->cfg, klass);
+
+				if (is_simd) {
+					g_assert (lhs);
+					int align = mono_class_value_size (klass, NULL);
+					retval = build_alloca_llvm_type_name (ctx, ret_type, align, "struct_ret");
+					retval = LLVMBuildLoad2 (builder, ret_type, retval, "struct_ret");
+
+					int len;
+					if (!strcmp ("Vector4", m_class_get_name (klass)))
```

Comment:
It might be better to use LLVMGetVectorSize (ret_type) to obtain the number of elements.

Reply:
Thanks for suggesting this. Fixed now.